### ISSUE TITLE
MixedLM nonconvergence

### ISSUE THREAD

**Issue by epierson9** [statsmodels/statsmodels#6157]
I'm using sm.MixedLM to fit quite a simple model: y ~ x, with group intercepts. (So the code I'm calling is just `model = sm.MixedLM.from_formula("y ~ x", df, groups=df["group"])` and then `result = model.fit(maxiter=30, start_params=start_params)`). x and y are continuous and roughly on the same scale. I'm using Python 3.7.2 and statsmodels 0.9.0. 

I'm fitting this model on many related datasets, and on some datasets it fails to converge with warnings like: 

```
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2027: ConvergenceWarning: Gradient optimization failed.
  warnings.warn(msg, ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2046: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2422: RuntimeWarning: invalid value encountered in sqrt
  sdf[0:self.k_fe, 1] = np.sqrt(np.diag(self.cov_params()[0:self.k_fe]))
/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py:877: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py:877: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py:1748: RuntimeWarning: invalid value encountered in greater_equal
  cond2 = (x >= self.b) & cond0
/lib/python3.7/site-packages/statsmodels/base/model.py:1092: RuntimeWarning: invalid value encountered in sqrt
  bse_ = np.sqrt(np.diag(self.cov_params()))
```

or, to take another example, 

```
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2027: ConvergenceWarning: Gradient optimization failed.
  warnings.warn(msg, ConvergenceWarning)
/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2046: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)
```

I'm not sure how to fix this. I've tried using a better initialization, which was somewhat helpful. I have a couple hundred to a couple hundred observations in each dataset, and an average of say 5 observations per group. It seems more likely to fail to converge on datasets where a) I have fewer observations or b) the estimated slope/Group Var parameters are smaller. Any suggestions you have for what is going wrong or what I should do are very welcome.

**Comment by kshedden** [comment]
These are warnings that we show as a way of alerting you that you may be in a non-standard situation.  Most likely, one of your variance parameters is converging to zero.  If you get a model result and the parameter estimates and standard errors are finite, I think you can just use them and ignore all the warnings.  Note that inference for variance parameters is not meaningful when the point estimate is at or near zero.

**Comment by kshedden** [comment]
It's a good idea to check the gradient, something like this should work:

```
model.score(result.params_object)
```

**Comment by epierson9** [comment]
Great, thanks so much for the quick reply -- Very helpful. Two followups -- 

1. "Note that inference for variance parameters is not meaningful when the point estimate is at or near zero" ~ do you mean that the standard errors on Group Var are not meaningful? (eg, 0.028 in the last row here --)
```
----------------------------------------------------------------
              Coef.   Std.Err.     z     P>|z|   [0.025   0.975]
----------------------------------------------------------------
Intercept     0.143      0.046   3.134   0.002    0.054    0.232
x             0.099      0.048   2.041   0.041    0.004    0.193
Group Var     0.085      0.028
==============================================================
```
2. Regarding `model.score(result.params_object)` -- yes, I am already storing this. How do I assess whether it has an acceptable value?

**Comment by kshedden** [comment]
Some people don't believe that the standard errors for variance parameters in mixed models are meaningful under any circumstances.  In R, they are not reported in the standard output.  The issue is that the MLE (or RMLE) converges very slowly to the limiting normal distribution, and if the true variance parameter is zero, the limiting distribution is not normal.  We opted to report these approximate (and sometimes invalid) standard errors here, and let the user decide what to do.   In most cases, the standard errors for variance and covariance parameters (of random effects) are not really needed.  But in the rare cases where they are, the usual advice is to use some form of simulation.  All that said, if your sample size is not very small, I don't consider 0.085 to be a particularly small variance parameter.   

However, given that your estimated variance parameter is not that small, now I question whether this is the root cause of all your warning messages.  Therefore, the score check is very important.  I might check the norm of the score vector, hopefully it is on the order of 10^-2 or smaller, but maybe 10^-1 is OK in some cases.  There isn't really a principled way to answer this because it depends on the scaling of the data and other factors.  You probably should drop the `maxiter=30` here, you want it to try to converge as much as possible.  You might also try other algorithms, like `method="bfgs"` or  `method="cg"` (as keyword arguments to `fit`).

**Comment by epierson9** [comment]
Oh, to be clear -- the model I pasted above converged. Sorry for the lack of clarity. I was just trying to make sure I knew which parameters you were referring to. Models that don't converge seem to have smaller slope or variance parameters, eg: 

```
----------------------------------------------------------------
            Coef.    Std.Err.     z      P>|z|   [0.025   0.975]
----------------------------------------------------------------
Intercept    0.690      0.008   89.718   0.000    0.675    0.705
x           -0.012      0.008   -1.582   0.114   -0.027    0.003
Group Var    0.005      0.002
==============================================================

----------------------------------------------------------------
             Coef.   Std.Err.     z      P>|z|   [0.025   0.975]
----------------------------------------------------------------
Intercept    0.281      0.009   31.605   0.000    0.263    0.298
x            0.022      0.009    2.379   0.017    0.004    0.041
Group Var    0.008      0.003
==============================================================
```
although that isn't always true: 
```
==============================================================
Model:            MixedLM Dependent Variable: y
No. Observations: 1159    Method:             REML
No. Groups:       126     Scale:              0.6624
Min. group size:  1       Likelihood:         -1450.9405
Max. group size:  143     Converged:          No
Mean group size:  9.2
----------------------------------------------------------------
             Coef.   Std.Err.     z      P>|z|   [0.025   0.975]
----------------------------------------------------------------
Intercept    0.586      0.046   12.688   0.000    0.495    0.676
x            0.072      0.047    1.518   0.129   -0.021    0.164
Group Var    0.113
==============================================================
```

I will make the other changes you suggest -- try other solvers, and increase maxiter. What do you recommend for a sufficient condition for when I have results "I can trust"? Based on your comments it seems like I can trust the results if: 

1. Converged is true, or
2. The norm of `model.score(result.params_object)` is < 10^-1 (How do I assess the scale of this parameter?), or
3. The variance parameter is very small with finite standard error. 

Does that seem reasonable?

**Comment by kshedden** [comment]
I'm not sure that there are explicit and universal criteria to use here.  Note that this issue is not unique to Statsmodels, most numeric optimization uses heuristic criteria for assessing numeric convergence.  Our results are extensively tested against R and generally agree well.  But R is a but less prone to producing convergence warnings.

Our warnings are particularly verbose in the setting where a variance parameter is converging to zero.  In this case, you could omit that variance parameter from the model and re-run it.  If you get no warnings when the parameter is omitted, you are probably fine.

The "Newton step" for optimizing a smooth function is inv(H) * G, where H is the Hessian and G is the gradient (score).  If this vector is small relative to the scale of the problem, you can generally treat the algorithm as converged.  A simpler way to go is just to look at G alone, without scaling by inv(H).  This is what I suggested before.  But this is more sensitive to the scale of the data.

We use scipy optimizers and generally trust the optimizers to decide if they are converged.  I assume they are well-tested and robust.  However for reasons I am not very clear on, sometimes one method works better than another, and there doesn't seem to be an obvious default optimizer that works best in every problem.

**Comment by epierson9** [comment]
Hi! Thank you for this detailed answer. We have followed up on this and are experimenting with the R implementation to see if it yields the same results as statsmodels while returning fewer convergence warnings. Our problem with statsmodels is that it's hard to assess whether we should be concerned about the frequent statsmodels warnings without a fair bit of expertise -- it seems from your answer like there are no hard rules. At least from our standpoint, the package would be more usable if it either returned fewer convergence warnings or if it were clearer when we should actually be concerned about them. In general I've found statsmodels very helpful, so thanks again for all your work developing it and for your thoughtful answers!
